File: packages/react-native-nitro-buffered-blob/android/src/main/java/com/margelo/nitro/bufferedblob/HybridNativeFileReader.kt
package com.margelo.nitro.bufferedblob

import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import java.io.FileInputStream
import java.io.File

@DoNotStrip
class HybridNativeFileReader(
  private val filePath: String,
  private val bufferSize: Int
) : HybridNativeFileReaderSpec() {

  private val file = File(filePath)
  private var inputStream: FileInputStream? = null
  @Volatile private var _bytesRead: Long = 0
  @Volatile private var _isEOF: Boolean = false
  @Volatile private var isClosed: Boolean = false

  override val memorySize: Long
    get() = bufferSize.toLong()

  override val fileSize: Long
    get() = file.length()

  override val bytesRead: Long
    get() = _bytesRead

  override val isEOF: Boolean
    get() = _isEOF

  init {
    if (!file.exists()) {
      throw Exception("[FILE_NOT_FOUND] File does not exist: $filePath")
    }
    if (!file.isFile) {
      throw Exception("[INVALID_ARGUMENT] Path is not a file: $filePath")
    }
    inputStream = FileInputStream(file)
  }

  override fun readNextChunk(): Promise<ArrayBuffer?> {
    return Promise.async(CoroutineScope(Dispatchers.IO)) {
      synchronized(this) {
        if (isClosed) {
          throw Exception("[STREAM_CLOSED] FileReader already closed: $filePath")
        }
        if (_isEOF) {
          return@async null
        }

        val stream = inputStream ?: throw Exception("[STREAM_CLOSED] FileReader already closed: $filePath")
        val buffer = ByteArray(bufferSize)
        val bytesReadNow = stream.read(buffer)

        if (bytesReadNow == -1) {
          _isEOF = true
          return@async null
        }

        _bytesRead += bytesReadNow

        val result = ArrayBuffer.allocate(bytesReadNow)
        result.getBuffer(false).put(buffer, 0, bytesReadNow)

        return@async result
      }
    }
  }

  override fun close() {
    synchronized(this) {
      if (isClosed) {
        return
      }
      isClosed = true
      try {
        inputStream?.close()
      } catch (e: Exception) {
        // Ignore close errors
      } finally {
        inputStream = null
      }
    }
  }
}

File: packages/react-native-nitro-buffered-blob/ios/HybridNativeFileReader.swift
import NitroModules
import Foundation

class HybridNativeFileReader: HybridNativeFileReaderSpec {
  private let inputStream: InputStream
  private let bufferSize: Int
  private var _bytesRead: Int64 = 0
  private let _fileSize: Int64
  private var _isEOF: Bool = false
  private var _isClosed: Bool = false
  private let queue: DispatchQueue

  var fileSize: Int64 {
    return _fileSize
  }

  var bytesRead: Int64 {
    return queue.sync { _bytesRead }
  }

  var isEOF: Bool {
    return queue.sync { _isEOF }
  }

  private var isClosed: Bool {
    get { queue.sync { _isClosed } }
    set { queue.sync { _isClosed = newValue } }
  }

  override var memorySize: Int {
    return bufferSize + MemoryLayout<HybridNativeFileReader>.size
  }

  init(path: String, bufferSize: Int, fileSize: Int64) throws {
    guard let stream = InputStream(fileAtPath: path) else {
      throw NSError(
        domain: "BufferedBlob",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "[FILE_NOT_FOUND] Could not open file: \(path)"]
      )
    }

    self.inputStream = stream
    self.bufferSize = bufferSize
    self._fileSize = fileSize
    self.queue = DispatchQueue(label: "com.bufferedblob.reader", qos: .userInitiated)

    super.init()

    stream.open()
    if stream.streamStatus == .error {
      throw NSError(
        domain: "BufferedBlob",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "[FILE_NOT_FOUND] Failed to open stream: \(path)"]
      )
    }
  }

  func readNextChunk() throws -> Promise<ArrayBuffer?> {
    if isClosed {
      throw NSError(
        domain: "BufferedBlob",
        code: 2,
        userInfo: [NSLocalizedDescriptionKey: "[INVALID_STATE] Reader is closed"]
      )
    }

    if isEOF {
      return Promise.resolved(with: nil)
    }

    return Promise.async { [weak self] in
      guard let self = self else {
        throw NSError(
          domain: "BufferedBlob",
          code: 3,
          userInfo: [NSLocalizedDescriptionKey: "[INVALID_STATE] Reader deallocated"]
        )
      }

      return try await withCheckedThrowingContinuation { continuation in
        self.queue.async {
          do {
            let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: self.bufferSize)
            defer { buffer.deallocate() }

            let bytesRead = self.inputStream.read(buffer, maxLength: self.bufferSize)

            if bytesRead < 0 {
              self._isEOF = true
              continuation.resume(returning: nil)
              return
            }

            if bytesRead == 0 {
              self._isEOF = true
              continuation.resume(returning: nil)
              return
            }

            self._bytesRead += Int64(bytesRead)

            let data = Data(bytes: buffer, count: bytesRead)
            let arrayBuffer = ArrayBuffer(data: data)

            continuation.resume(returning: arrayBuffer)
          } catch {
            continuation.resume(throwing: error)
          }
        }
      }
    }
  }

  func close() throws {
    if isClosed {
      return
    }

    isClosed = true
    inputStream.close()
  }

  deinit {
    try? close()
  }
}
